### DISCUSSION TITLE
Retrieve data from a page in another page efficiently

### DISCUSSION THREAD

**Discussion by sarahdayan** [vercel/next.js#20243] [Help]
Hi! I’m exploring Next.js to see if it’s fit to replace an aging Ruby stack, and trying to come up with solutions to substitute the important features.

One thing we often do in the project is **query data from a page in another page.** For example, we have a component which takes an array of internal URLs as an input. It retrieves the matching page’s titles from their frontmatter, and outputs them as a list of links. We’re able to do this efficiently on thousands of pages because every page in the project is loaded in memory when starting the project, and accessible in a tree.

I couldn’t find detailed documentation on this but I don’t think Next.js has a similar data structure available, or at least not exposed for consumption. Currently, the one naive approach I can think of to replicate this behavior with JSX components in MDX files is to write a remark or rehype plugin that grabs the passed URLs and reads each matching file from the file system to retrieve the title. I fear that it might have really poor performance though, and I’m wondering if there’s a better way.

Any available option I’m not thinking of?

Thanks in advance for your help!

**Comment by saleebm** [comment]
I think I also asked the same question [here in my discussion post](https://github.com/vercel/next.js/discussions/20237#discussion-73162). I have the same concern. With Gatsby, it would be possible to useStaticQuery in that component. 

An option is to compose getStaticProps with the common data requirements and share it in the pages. Although I think a better solution would be a hook or app props, then work out the implementation of the static data obtained at the root from the server-rendered pages (if any).

Another option is to use getInitialProps in the app for the blocking data requirements, but that would force your entire app to be ssr.

**Reply by sarahdayan** [reply]
Thanks @saleebm for your reply. In my case, the data would be fetched from MDX content files, so I'm not sure `getStaticProps` is accessible from there.

All routes of the site could potentially be called so I'm looking for a solution that would let me access any of them. Indeed, the `useStaticQuery` approach from Gatsby would help here, but AFAIK Next.js doesn't have this concept of a centralized data store, at least not publicly.
